### PR TITLE
Category sync

### PR DESCRIPTION
**Pull request by hajoeichler**
With this PR we are able to update categories to any supported update action - currently there are 6 actions.

@butenkor and @emmenko  please review.

### CODE REVIEW

File: src/coffee/sync/category-sync.coffee [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+###*
+ * CategorySync Sync class
+###
+class CategorySync extends BaseSync
+
+  constructor: ->
+    # Override base utils
+    @_utils = new CategoryUtil()
+
+  _doMapActions: (diff, new_obj, old_obj) ->
```
Comment (butenkor) [inline src/coffee/sync/category-sync.coffee]: is old_obj param required here?

File: src/coffee/sync/utils/category-utils.coffee [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+class CategoryUtils extends BaseUtils
+
+  ###
+  Create list of actions for syncing categories.
+  @param {object} diff result of jsondiffpatch tool.
+  @return list with actions
+  ###
+  actionsMap: (diff, new_obj) ->
+    actions = []
+    if diff?
```
Comment (butenkor) [inline src/coffee/sync/utils/category-utils.coffee]: isn't just ```if diff``` not enough?

File: src/coffee/sync/category-sync.coffee (hunk 1 above)
Comment (emmenko) [inline src/coffee/sync/category-sync.coffee]: Is not required, but this is the method definition. It's up to the actual utility functions to use it or not

File: src/spec/sync/utils/category-utils.spec.coffee [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+      otherCategory = _.deepClone category
+      otherCategory.externalId = 'ext-123'
+
+      delta = @utils.diff category, otherCategory
+      update = @utils.actionsMap delta, otherCategory
+      expect(update).toEqual [
+        { action: 'setExternalId', externalId: 'ext-123' }
+      ]
+
+    it 'should create action to set external id', ->
```
Comment (emmenko) [inline src/spec/sync/utils/category-utils.spec.coffee]: Duplicate test?

File: src/spec/sync/category-sync.spec.coffee [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+  beforeEach ->
+    @sync = new CategorySync
+
+  afterEach ->
+    @sync = null
+
+  describe ':: buildActions', ->
+
+    it 'no differences', ->
```
Comment (emmenko) [inline src/spec/sync/category-sync.spec.coffee]: `should not create payload`

File: src/spec/sync/category-sync.spec.coffee [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+      category =
+        id: 'c123'
+
+      update = @sync.buildActions(category, category).getUpdatePayload()
+      expect(update).toBeUndefined()
+      updateId = @sync.buildActions(category, category).getUpdateId()
+      expect(updateId).toBe 'c123'
+
+    it 'various differences', ->
```
Comment (emmenko) [inline src/spec/sync/category-sync.spec.coffee]: `should create different type of actions`

File: src/coffee/sync/utils/category-utils.coffee (hunk 2 above)
Comment (emmenko) [inline src/coffee/sync/utils/category-utils.coffee]: Yes

File: src/spec/sync/category-sync.spec.coffee (hunk 4 above)
Comment (butenkor) [inline src/spec/sync/category-sync.spec.coffee]: better ```should not build update actions```?

**Comment by emmenko** [comment]
Looks good, clean and simple ;)

Look at my comments, then you can merge :+1:

File: src/coffee/sync/utils/category-utils.coffee [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+      _.each actionsList(), (item) =>
+        key = item.key
+        obj = diff[key]
+        if obj?
```
Comment (butenkor) [inline src/coffee/sync/utils/category-utils.coffee]: Then same here

**Comment by butenkor** [comment]
So sync will handle category updates only? Creates and deletes will be handled by https://github.com/sphereio/sphere-category-sync?

File: src/spec/sync/utils/category-utils.spec.coffee (hunk 3 above)
Comment (hajoeichler) [inline src/spec/sync/utils/category-utils.spec.coffee]: Will adjust to `change external id`. Was late ;)

File: src/spec/sync/category-sync.spec.coffee (hunk 4 above)
Comment (hajoeichler) [inline src/spec/sync/category-sync.spec.coffee]: Yep, will adapt to a (much) more meaningful description.

**Pull request merged**